Okay, I'm having trouble with this because the fact that using an `llvmcall()` intrinsic causes an error the first time keeps on breaking my bootstrap, even if I wrap it in a `try/catch` and put it at the end of the bootstrap right before calling `precompile.jl`.  Specifically, this is the code I'm moving around:

```
# fused multiply-add
fma_libm(x::Float32, y::Float32, z::Float32) =
    ccall(("fmaf", libm_name), Float32, (Float32,Float32,Float32), x, y, z)
fma_libm(x::Float64, y::Float64, z::Float64) =
    ccall(("fma", libm_name), Float64, (Float64,Float64,Float64), x, y, z)
fma_llvm(x::Float32, y::Float32, z::Float32) =
    box(Float32,fma_float(unbox(Float32,x),unbox(Float32,y),unbox(Float32,z)))
fma_llvm(x::Float64, y::Float64, z::Float64) =
    box(Float64,fma_float(unbox(Float64,x),unbox(Float64,y),unbox(Float64,z)))
# Disable LLVM's fma if it is incorrect, e.g. because LLVM falls back
# onto a broken system libm; if so, use openlibm's fma instead. Also
# fallback to opeblibm if it messes with the floating point rounding
# 1.0000305f0 = 1 + 1/2^15
# 1.0000000009313226 = 1 + 1/2^30

function get_mxcsr_rounding()
    mxcsr = llvmcall( """%ptr = alloca i32
              call void @llvm.x86.sse.stmxcsr(i32 * %ptr)
              %curval = load i32 * %ptr
              ret i32 %curval""", UInt32, () )
    bits(mxcsr)[18:19]
end

# Need to warmup llvmcall due to https://github.com/JuliaLang/julia/pull/8740
try
    get_mxcsr_rounding()
end
initial_rounding = get_mxcsr_rounding()

if (fma_llvm(1.0000305f0, 1.0000305f0, -1.0f0) == 6.103609f-5 &&
    (fma_llvm(1.0000000009313226, 1.0000000009313226, -1.0) ==
     1.8626451500983188e-9) && initial_rounding == get_mxcsr_rounding() )
    fma(x::Float32, y::Float32, z::Float32) = fma_llvm(x,y,z)
    fma(x::Float64, y::Float64, z::Float64) = fma_llvm(x,y,z)
else
    fma(x::Float32, y::Float32, z::Float32) = fma_libm(x,y,z)
    fma(x::Float64, y::Float64, z::Float64) = fma_libm(x,y,z)
end
# This is necessary at least on 32-bit Intel Linux, since fma_llvm may
# have called glibc, and some broken glibc fma implementations don't
# properly restore the rounding mode
Rounding.set_rounding_raw(Float32, Rounding.JL_FE_TONEAREST)
Rounding.set_rounding_raw(Float64, Rounding.JL_FE_TONEAREST)
```

After bootstrapping, I can run this code no problem and I get the desired result, but in the middle of bootstrapping, I invariably get an error like the following:
```
llvmintrinsics.jl
error during bootstrap: LoadError(at "sysimg.jl" line 297: LoadError(at "llvmintrinsics.jl" line 33: ErrorException("error compiling anonymous: Failed to parse LLVM Assembly:
julia: <string>:4:11: error: use of undefined value '@llvm.x86.sse.stmxcsr'
call void @llvm.x86.sse.stmxcsr(i32 * %ptr)
          ^
")))
Basic Block in function 'julia_anonymous_17814' does not have terminator!
label %try
LLVM ERROR: Broken module, no Basic Block terminator!
make[1]: *** [/home/vagrant/buildbot/slave/build_ubuntu14_04-x86/build/usr/lib/julia/sys0.o] Error 1
make: *** [release] Error 2
```

Where `llvmintrinsics.jl` is a file I added to the system image generation process at the end to avoid any issues with not having things like `@mstr` defined.  The backtrace here is for running the slightly modified code:
```
try
    llvmcall( """%ptr = alloca i32
              call void @llvm.x86.sse.stmxcsr(i32 * %ptr)
              %curval = load i32 * %ptr
              ret i32 %curval""", UInt32, () )
end
```
as my warm-up step.  Any ideas?

--- a/base/float.jl
+++ b/base/float.jl
@@ -397,18 +397,23 @@ fma_llvm(x::Float64, y::Float64, z::Float64) =
 # Disable LLVM's fma if it is incorrect, e.g. because LLVM falls back
 # onto a broken system libm; if so, use openlibm's fma instead
 # 1.0000305f0 = 1 + 1/2^15
-if fma_llvm(1.0000305f0, 1.0000305f0, -1.0f0) == 6.103609f-5
-    fma(x::Float32, y::Float32, z::Float32) = fma_llvm(x,y,z)
-else
-    fma(x::Float32, y::Float32, z::Float32) = fma_libm(x,y,z)
-end
 # 1.0000000009313226 = 1 + 1/2^30
-if (fma_llvm(1.0000000009313226, 1.0000000009313226, -1.0) ==
-    1.8626451500983188e-9)
+# If fma_llvm() clobbers the rounding mode, the result of 0.1 + 0.2 will be 0.3
+# instead of the properly-rounded 0.30000000000000004; check after calling fma
+if (fma_llvm(1.0000305f0, 1.0000305f0, -1.0f0) == 6.103609f-5 &&
+    (fma_llvm(1.0000000009313226, 1.0000000009313226, -1.0) ==
+     1.8626451500983188e-9) && 0.1 + 0.2 == 0.30000000000000004)
+    fma(x::Float32, y::Float32, z::Float32) = fma_llvm(x,y,z)
     fma(x::Float64, y::Float64, z::Float64) = fma_llvm(x,y,z)
 else
+    fma(x::Float32, y::Float32, z::Float32) = fma_libm(x,y,z)
     fma(x::Float64, y::Float64, z::Float64) = fma_libm(x,y,z)
 end
+# This is necessary at least on 32-bit Intel Linux, since fma_llvm may
+# have called glibc, and some broken glibc fma implementations don't
+# properly restore the rounding mode
+Rounding.set_rounding_raw(Float32, Rounding.JL_FE_TONEAREST)
+Rounding.set_rounding_raw(Float64, Rounding.JL_FE_TONEAREST)
 
 ## byte order swaps for arbitrary-endianness serialization/deserialization ##
 bswap(x::Float32) = box(Float32,bswap_int(unbox(Float32,x)))

